### ISSUE TITLE
Question: How do I distribute a calmjs-enabled Python package, ensuring that the "user" has nodejs installed and can run the JavaScript code?

### ISSUE THREAD

**Issue by s-m-e** [calmjs/calmjs#40]
@metatoaster Thanks for this really excellent package. 

I *believe* I understand how I define JavaScript packages, which my Python code depends on, with the help of calmjs. What I really do NOT understand however is how one ensures that there is a (recent / decent, working) installation of node.js on the "user's" system and how it is accessed by a calmjs-enabled Python package. I guess an average Python user does not expect a JavaScript dependency, requiring another interpreter. Is there a way to address this issue with calmjs?

---

I have a few Python packages with JavaScript dependencies currently in local use, which I'd like to publish. I am working with the following hack, either in a Bash script or makefile, for installing my dependencies and ensuring that nodejs is there:

```shell
cd /root/of/python/package
pip install -r requirements_python.txt # includes "nodeenv"
nodeenv -p # pulls node.js and integrates it into my virtual environment
npm i -g npm # update npm ...
cat requirements_node.txt | xargs npm install -g
pip install -e . # remove "-e" for deployment
```

I have two requirements files, one for Python and one for node.js. The most critical step though is `nodeenv -p`, which pulls a pre-build node binary into the current (activated) virtual environment. This approach has a number of more or less significant drawbacks, so I am looking for alternatives.

I posted a [similar question on SO](https://stackoverflow.com/q/49953785/1672565).

**Comment by s-m-e** [comment]
@metatoaster Thanks a lot for your answer and sorry it took me a while to revisit this issue. You partially answer my question or at least you let me understand where I have to dig deeper. If I understand you correctly, `calmjs` helps me to define JavaScript / node dependencies but explicitly does not take care of "plowing the ground", i.e. ensuring that node and/or npm is installed and configured in a sane manner.

What I had in mind can be summarized as follows. For development, I heavily rely on the `nodeenv` Python package. I assume you're familiar with it (?). If you're in a Python virtual environment, you can simply run `nodeenv -p` and `nodeenv` will (a) load a node binary (matching your architecture and operating system etc), (b) install it into your virtual environment's directory tree and (c) add configuration options to the `activate` script of your virtual environment. Once this is done (fully automatically), you can install node packages with `npm install -g ...` "globally" into your virtual environment. npm's "global scope" matches the scope of your Python virtual environment. So, `nodeenv` really does a lot of the hard work of allowing you to run node, regardless of your operating system or architecture. That's the upside.

The downside is that I need to tell my users that they need a virtual environment. I also need to tell them that it should be clean or at least not have `nodeenv` and node in it already. One of `nodeenv`'s really cool features could potentially be problematic here: As an alternative to a full node build, you can also pull (or even compile?) partial or custom configured versions of node. So running something like `nodeenv -p` in a user's virtual environment could have potentially unwanted side-effects if there was a pre-existing deployment of node. All of that could be an issue / feature request against `nodeenv`, true, but there is also another way of looking at it.

Imagine `calmjs` has some of the capabilities of `nodeenv` (or would actually just use `nodeenv` as a dependency). Instead of applying changes to the entire virtual environment during installation, I could imagine that a `calmjs`-enabled package could (optionally) with the help of `calmjs` pull node and npm into its package folder (in `site-packages`) during installation and configure it in a sane way. The package folder could be the "root" of npm's "global scope" just for this one single Python package. If JavaScript code is then called from this Python package through `calmjs`, `calmjs` would know where to look for node, the script and its dependencies.  

I hope I got my idea across. What are your thoughts on that?

**Comment by metatoaster** [comment]
> If I understand you correctly, calmjs helps me to define JavaScript / node dependencies but explicitly does not take care of "plowing the ground", i.e. ensuring that node and/or npm is installed and configured in a sane manner.

Yes, essentially that.  Likewise, `calmjs` doesn't (actually cannot) ensure that the Python is installed correctly. There are system level package managers that are written in Python that can manage the installation of Node.js/npm and their dependencies (C/C++ compilers, glibc, etc), but that's usually done by organisations that build distributions.  I've done something similar where I leveraged an existing distribution and virtualisation platforms (VirtualBox in my case) and orchestrate an installation of multiple platforms all at once, including Python, Java/Tomcat and others.  If I needed `calmjs` for that (eventually it will be) I can specify that in the Python specific dependencies for that.

I guess I can expand on this more on this not exactly relevant but still tangentially relevant subject. Specifically, it's not so much a dependency manage, but it's more an orchestration script for building `.vmdk` or `.ova` for distribution to end-users and/or cloud providers, but it does use the Linux distribution's binary packages and/or set up the environment to build the application required. The repository is called [vboxtools](https://github.com/metatoaster/vboxtools) but it's barely alpha quality and it only satisfy some internal organisational requirements (quick way to reproduce complete builds/test live production backup/restore and provisioning new development/staging/production VMs), but it gets the job done to the most pedantic level possible, i.e. full control of a fully functional operating system that users can run using their native VirtualBox client, but the downside is that it's even heavier than distributing binaries for all supported platforms.  Only really suitable for deployment to cloud or passing around pre-installed USB sticks in a conference room/workshop.

> What I had in mind can be summarized as follows.

Essentially that.  I mean, for development I tell my fellow developers to make use of Python's `virtualenv` (or like it) to set up their environments for the required Python versions, and then `nodeenv` would fit right into that.

Now with `calmjs`, we can also do `calmjs npm --install <pkg>` or if we are running a `setup.py`, after running `python setup.py develop` to get the package installed using development mode, we can also do `python setup.py npm --install` to install the dependencies defined in `package_json` section of `setup.py` for that and their Python `install_requires` dependencies.  Developers should be able to do this.

For users, we have to ask: which users (also on what OS platform, getting to this later)?  They come in all forms (biological/digital), and have different skill levels (including developers, which we just covered).  From the sound of it you are looking at potentially minimally technical users that might be able to fumble around on the command line, but may not necessary know what went wrong if the commands they are typing isn't doing what was supposed to happen.  Well, one way you can work around this to actually have a build script that will download/fetch all the dependencies (like downloading/compiling Python if it can't find one, downloading/compiling Node.js if it can't find one, etc) to produce a working environment on the local directory (on standard POSIX environment, it shouldn't conflict with system level stuff because all of that is isolated in the local directory, provided that the `./configure` scripts have the right paths passed to it, for instance.  So there isn't exactly anything wrong with the `nodeenv` approach, just that it is better to have some script (like `bash` and/or `batch`/`powershell` if Windows is needed) that will download/install all runtime dependencies.

> Imagine `calmjs` has some of the capabilities of `nodeenv`

Nope, not likely to happen.  Even though it is possible to declare Node.js version requirements for a given Node.js package, `npm` itself doesn't go upgrade Node.js for you - it will just merely tell the user to install the right version.  But what I described earlier (e.g. some form of shell scripting using the operating system's scripting environment) may be a way to ensure that the environment is sane for the package level installation step to continue, and if the requirements cannot be satisfied it might go and try downloading/installing those binary requirements.  Then `pip` or `npm` will become available and so the installation of packages from those repositories can be done.

> If JavaScript code is then called from this Python package through calmjs, calmjs would know where to look for node, the script and its dependencies.

Yes, packages like [`calmjs.webpack`](https://github.com/calmjs/calmjs.webpack), [`calmjs.rjs`](https://github.com/calmjs/calmjs.rjs) and [`calmjs.dev`](https://github.com/calmjs/calmjs.dev) (the `karma` test runner integration) (all of which are shameless plugs) make use of the `calmjs.toolchain` module which provide some very specific form of integration so it might not be applicable for the general use case. However the simple one that implements the `calmjs npm` tool effectively extends upon the basic [`BaseDriver` class](https://github.com/calmjs/calmjs/blob/3.0.0/src/calmjs/base.py#L363) to locate the `node_modules` binary.  That said, if you are going to do the effort to have complete builds of the your application stack, you don't even need that when your environment will have all the Node.js "binaries" (Node.js scripts in `./node_modules/.bin`) in a known location and you can just have Python running them directly.

Again, this is a lot to take in, so I don't blame you for taking a long time to come back to this.  As an aside, it took me a good few months of working on the original progeny project that started me thinking about this issue before it got me to spend three months to come up with a minimum viable package, then I found myself spending the next two years polishing it to where I can maybe get people to look at it.  So yes, this is a truly a hard problem, and in some ways an artificial one which people have said will go away with WebAssembly, but if the development is not web related it doesn't actually help either (hence I focused `calmjs` on helping out with getting Node.js dependencies tracked in Python).

**Comment by s-m-e** [comment]
@metatoaster Thanks for all the effort and your again well informed answer. I am still chewing on it ;) Meanwhile, I opened a [question in the nodeenv issue tracker](https://github.com/ekalinin/nodeenv/issues/214), asking how I could make a Python package depend on `nodeenv`. The combination of both your package and `nodeenv` could be a path providing something like a viable solution, I hope ...
